gadget/install/partition.go

```diff
@@ -69,6 +84,103 @@ func createMissingPartitions(dl *gadget.OnDiskVolume, pv *gadget.LaidOutVolume)
 	return created, nil
 }
 
+// buildPartitionList builds a list of partitions based on the current
+// device contents and gadget structure list, in sfdisk dump format, and
+// returns a partitioning description suitable for sfdisk input and a
+// list of the partitions to be created.
+func buildPartitionList(dl *gadget.OnDiskVolume, pv *gadget.LaidOutVolume) (sfdiskInput *bytes.Buffer, toBeCreated []gadget.OnDiskStructure) {
+	sectorSize := dl.SectorSize
+
+	// Keep track what partitions we already have on disk
+	seen := map[quantity.Offset]bool{}
+	for _, s := range dl.Structure {
+		start := s.StartOffset / quantity.Offset(sectorSize)
+		seen[start] = true
+	}
+
+	// Check if the last partition has a system-data role
+	canExpandData := false
+	if n := len(pv.LaidOutStructure); n > 0 {
+		last := pv.LaidOutStructure[n-1]
+		if last.VolumeStructure.Role == gadget.SystemData {
+			canExpandData = true
+		}
+	}
+
+	// The partition index
+	pIndex := 0
+
+	// Write new partition data in named-fields format
+	buf := &bytes.Buffer{}
+	for _, p := range pv.LaidOutStructure {
+		if !p.IsPartition() {
+			continue
+		}
+
+		pIndex++
+		s := p.VolumeStructure
+
+		// Skip partitions that are already in the volume
+		start := p.StartOffset / quantity.Offset(sectorSize)
+		if seen[start] {
+			continue
+		}
+
+		// Only allow the creation of partitions with known GUIDs
+		// TODO:UC20: also provide a mechanism for MBR (RPi)
```

Comment:
this todo can go away I think

Reply:
I'm not sure, the issue is that the next line check starts with dl.Schema == "gpt" ? afaict this is unrelated to the work we did for  wasCreatedDuringInstall ?

Reply:
well the issue is that we don't have GUID's on MBR to check to see if we can create/re-create, but we do have sufficient code to create the right partitions on MBR, so I'm not sure what we have left to do here.
In other words, if this TODO wasn't addressed then we wouldn't be able to install on a pi which uses MBR. In case the TODO was about re-creating partitions, I did just confirm that we can reinstall properly on my pi running uc20 and that worked fine, so I don't think that there are any TODO's left about creating/recreating partitions on MBR devices

Reply:
it's really about limiting the types to a known lists, we do that for gpt and don't for mbr. Is that super useful, not sure given we don't use the attribute bits anymore? maybe it should just become a XXX,

// XXX this safeguard check is done only for gpt

Otherwise the other approach would be to drop the check also for gpt? I don't think that would break something either at this point

Reply:
ah yes I see your point now, yes in that case then you are right we should probably just turn it into XXX and reword the comment, as I had misunderstood the code due to the comment

Reply:
but also yes we could potentially just drop the check entirely for GPT since we don't use that attr, I don't have strong opinions either way

Reply:
anyway thatTODO or code isn't new to this PR, it's only been moved around